### PR TITLE
Adding transaction id for waf log

### PR DESCRIPTION
**Pull request by azhao155**
Adding nginx request_id into waf log:
example
{
    "timeStamp": "2019-11-15T00:32:34+00:00",
    "resourceId": "resource",
    "operationName": "ApplicationGatewayFirewall",
    "category": "ApplicationGatewayFirewallLog",
    "properties": {
        "instanceId": "abc",
        "clientIp": "127.0.0.1",
        "clientPort": "",
        "requestUri": "/",
        "ruleSetType": "",
        "ruleSetVersion": "",
        "ruleId": "949110",
        "message": "Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total Score: 5)",
        "action": "Blocked",
        "site": "Global",
        "details": {
            "message": "Access denied with code 403 (phase 2). Operator GE matched 5 at TX:anomaly_score. ",
            "data": "",
            "file": "\"/root/CP/Nginx/crs/crs-3.0/rules/REQUEST-949-BLOCKING-EVALUATION.conf",
            "line": "57"
        },
        "hostname": "localhost:8080",
        "transactionId": "154826be7007cf6e6c363dc25ffb3866",
        "policyId": "",
        "policyScope": "",
        "policyScopeName": ""
    }
}
{
    "timeStamp": "2019-11-15T00:32:34+00:00",
    "resourceId": "resource",
    "operationName": "ApplicationGatewayFirewall",
    "category": "ApplicationGatewayFirewallLog",
    "properties": {
        "instanceId": "abc",
        "clientIp": "127.0.0.1",
        "clientPort": "",
        "requestUri": "/",
        "ruleSetType": "",
        "ruleSetVersion": "",
        "ruleId": "980130",
        "message": "Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0): SQL Injection Attack: SQL Tautology Detected.",
        "action": "Blocked",
        "site": "Global",
        "details": {
            "message": "Warning. Operator GE matched 5 at TX:inbound_anomaly_score. ",
            "data": "",
            "file": "\"/root/CP/Nginx/crs/crs-3.0/rules/RESPONSE-980-CORRELATION.conf",
            "line": "73"
        },
        "hostname": "localhost:8080",
        "transactionId": "154826be7007cf6e6c363dc25ffb3866",
        "policyId": "",
        "policyScope": "",                                    
}

### CODE REVIEW

**Review by allanrbo — DISMISSED** [review]

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 1]
```diff
@@ -352,6 +370,12 @@ ngx_http_modsecurity_load_request(ngx_http_request_t *r)
 
     req->header_only = r->header_only ? r->header_only : (r->method == NGX_HTTP_HEAD);
 
+#ifdef WAF_JSON_LOGGING_ENABLE
+    ngx_str_t  request_id;
+    request_id = get_request_id(r);
```
Comment (vkrivopalov) [inline nginx/modsecurity/ngx_http_modsecurity.c]: This two strings can be merged into one

File: apache2/apache2_util.c [hunk 2]
```diff
@@ -351,7 +350,7 @@ static void send_waf_log(struct waf_lock* lock, apr_file_t* fd, const char* str1
     char timestamp[30];
     strftime(timestamp, sizeof(timestamp), "%Y-%m-%dT%H:%M:%S+00:00", &timeinfo);
 
-    char* json_str = generate_json(timestamp, msc_waf_resourceId, WAF_LOG_UTIL_OPERATION_NAME, WAF_LOG_UTIL_CATEGORY, msc_waf_instanceId, waf_ip, waf_port, uri, waf_ruleset_type, waf_ruleset_version, waf_id, waf_message, mode, 0, waf_detail_message, waf_data, waf_filename, waf_line, hostname, waf_unique_id, waf_policy_id, waf_policy_scope, waf_policy_scope_name);
+    char* json_str = generate_json(timestamp, msc_waf_resourceId, WAF_LOG_UTIL_OPERATION_NAME, WAF_LOG_UTIL_CATEGORY, msc_waf_instanceId, waf_ip, waf_port, uri, waf_ruleset_type, waf_ruleset_version, waf_id, waf_message, mode, 0, waf_detail_message, waf_data, waf_filename, waf_line, hostname, unique_id, waf_policy_id, waf_policy_scope, waf_policy_scope_name);
```
Comment (vkrivopalov) [inline apache2/apache2_util.c]: The `unique_id` name is very generic and may be ambiguous. How about `request_unique_id`?

File: apache2/apache2_util.c [hunk 3]
```diff
@@ -488,7 +487,7 @@ static void internal_log_ex(request_rec *r, directory_config *dcfg, modsec_rec *
         const char* scope = apr_table_get(r->notes, WAF_POLICY_SCOPE);
         const char* scope_name = apr_table_get(r->notes, WAF_POLICY_SCOPE_NAME);
 
-        send_waf_log(msr->modsecurity->wafjsonlog_lock, msc_waf_log_fd, str1, r->useragent_ip ? r->useragent_ip : r->connection->client_ip, log_escape(msr->mp, r->uri), (!msr->allow_scope) ? dcfg->is_enabled : msr->allow_scope, r->hostname, unique_id, r, dcfg->waf_policy_id, scope ? scope : "", scope_name ? scope_name : "");
+        send_waf_log(msr->modsecurity->wafjsonlog_lock, msc_waf_log_fd, str1, r->useragent_ip ? r->useragent_ip : r->connection->client_ip, log_escape(msr->mp, r->uri), (!msr->allow_scope) ? dcfg->is_enabled : msr->allow_scope, r->hostname, r->log_id, r, dcfg->waf_policy_id, scope ? scope : "", scope_name ? scope_name : "");
```
Comment (vkrivopalov) [inline apache2/apache2_util.c:489]: Same here - it is not exactly clear what `log ID` means. Maybe use something like `request_id` instead?

**Review by vkrivopalov — COMMENTED** [review]
Looks fine overall, maybe some renaming will make it even better.

File: apache2/apache2_util.c (hunk 3 above)
Comment (azhao155) [reply apache2/apache2_util.c:489]: for the per request req_rec struct, it's defined by the apache lib, and it only has log_id for the for logging identify.

Comment (vkrivopalov) [reply apache2/apache2_util.c:489]: Fair enough, let's keep it as is then.

**Review by vkrivopalov — APPROVED** [review]
Looks good, thanks.

**Pull request merged**